ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -5839,13 +5840,30 @@ public HiveMetaHook getHook(
       }
     };
 
+    if ("rest".equals(conf.get("metastore.type", "hms"))) {
+      return getHiveIcebergRESTCatalog(conf);
+    }
+
     if (conf.getBoolVar(ConfVars.METASTORE_FASTPATH)) {
       return new SessionHiveMetaStoreClient(conf, hookLoader, allowEmbedded);
     } else {
       return RetryingMetaStoreClient.getProxy(conf, hookLoader, metaCallTimeMap,
           SessionHiveMetaStoreClient.class.getName(), allowEmbedded);
     }
   }
+  private IMetaStoreClient getHiveIcebergRESTCatalog(HiveConf conf) throws MetaException {
+    try {
+      Class<? extends IMetaStoreClient> handlerClass =
+          (Class<? extends IMetaStoreClient>)
+              Class.forName("org.apache.iceberg.hive.HiveIcebergRESTCatalogClientAdapter", true, Utilities.getSessionSpecifiedClassLoader());
+      IMetaStoreClient restCatalogMetastoreClient = ReflectionUtils.newInstance(handlerClass, conf);
+      restCatalogMetastoreClient.reconnect();
+      return restCatalogMetastoreClient;
```

Comment:
I don't disagree with this approach. However, if we replace IMetaStoreClient, I'd like you to resolve HIVE-12679, which is more generic and allows users to use Glue or another service.

Committers have asked us to discover solutions to implement some of SessionHiveMetaStoreClient's traits.
https://github.com/apache/hive/pull/4444

That's why we should make configurable not IMetaStoreClient but ThriftHiveMetastore.Iface.
https://issues.apache.org/jira/browse/HIVE-28658?focusedCommentId=17905019&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17905019

Reply:
@okumin thanks for checking my PR. I see the point to make it configurable, but as I see the ticket you mentioned didn't get a conclusion and it was closed and not merged. So I would suggest to handle this separately and once your is merged we can refactor this. WDYT?

Reply:
@okumin could you please reopen your PR?

Reply:
My PR has not reached any conclusion with which maintainers can agree. We have to reconsidering how we integrate some session features such as tmp tables.
This is the summary and my prososal. I'd like to hear your opinions as my recommendation is not aligned your PR

https://docs.google.com/document/d/1fFvB0DAXJvPYv27R8nLa3tOUT67-hY0owxBrDY-BxUA/edit?usp=sharing

Reply:
i think the time has come for HIVE-12679. We should start a new discussion and proceed with the approach with more +1 votes.

Reply:
@ngsg is PoCing the feasibility of a custom IMetaStoreClient before starting the discussion. Stay tuned
https://issues.apache.org/jira/browse/HIVE-27473

Reply:
Thank you all for bringing this topic! I think we should handle this once those tickets got merged. I am happy to refactor this after.